🎨 zb: Remove unsafe type casting code

sander-machado opened the pull request:
Rustc 1.86 stabilizes a new features called trait_upcasting. As quoted from the rust docs "Enable upcasts from dyn Trait1 to dyn Trait2 if Trait1 is a subtrait of Trait2". Which is exactly what this unsafe code was doing. I added the new syntax and increased the compiler version to 1.86 to enable this feature and remove these unsafe blocks. There were also multiple places where this check was called without any reason as it's checked at construction (and would just panic either way).

Related to: https://github.com/rust-lang/rust/issues/65991

<!--

Thank you for your contribution! 🙏

We hope you have read our contribution guideline and followed it to the best of your abilities:

https://github.com/dbus2/zbus/blob/main/CONTRIBUTING.md#submitting-pull-requests

-->

zeenix commented:
Cool. I'll look at it in more detail later but in the meantime, please split the commit to keep commits atomic. Here, I see at least 3 logical changes: removal of unneeded checks (which I'm not too sure about), bumping of MSRV and removal of unsafe code.

sander-machado commented:
> @sander-machado I think Rust 1.86 is not too new anymore. Let's bump the MSRV as part of this PR? You'll want to rebase though.

Great I'll update my PR once that is done!

sander-machado committed:
🎨 zb: Replace unsafe downcast API w/ std's

Rustc 1.86 stabilizes a new feature called trait_upcasting. As quoted from the rust docs "Enable upcasts from dyn Trait1 to dyn Trait2 if Trait1 is a subtrait of Trait2". Which is exactly what this unsafe code was doing. Added the new syntax to enable this feature and remove these unsafe blocks. There were also multiple places where this check was called without any reason as it's checked at construction (and would just panic either way).

Related to: rust-lang/rust#65991

sander-machado committed:
🎨 zb: Remove redundant downcast checks

These downcasts are already validated at construction time and would just panic either way, so they serve no purpose.

zeenix reviewed (changes requested):
LGTM apart from one thing and the usual nitpicks on commit messages: 😄 

* They're missing the `zb: ` prefix.
* second one could be shorter to be close to the guideline, something like`Replace unsafe downcast API w/ std's`.
* A description would be nice, even if a short one. For the second commit, you could use the same description as the PR's since this is the main commit.
* typo: build ins -> built-ins. However this becomes irrelevant if you change the title.

sander-machado commented on zbus/src/object_server/mod.rs:
```diff
@@ -272,11 +272,11 @@ impl ObjectServer {
+        {
+            // Ensure what we return can later be dowcasted safely.
+            let lock: &dyn Any = &*lock.read().await;
+            let _lock: &I = lock.downcast_ref::<I>().ok_or(Error::InterfaceNotFound)?;
```
you need to explicitly make the detour with the &dyn Any otherwise the compiler doesn't like it

zeenix replied:
Ah ok. Would be good to change the variable names still. 😉

zeenix replied:
@sander-machado could you please take this to the finish line? :pray:

zeenix reviewed (changes requested):
The changes look good to me, but please use 🦺 as the emoji prefix and ensure that lines in description of git commit message < 75 chars (as per our and git conventions).
